File: PYTHON/coast/coast.py
# Rating: ~ 3.3 / 10
# Link: https://open.kattis.com/problems/coast

from sys import stdin, stdout

def bfs(levels, r, c, visited):
  queue = [(0, 0)]
  visited[0][0] = True

  count = 0

  while queue:
    curr = queue.pop()
    row = curr[0]
    col = curr[1]

    if row - 1 > -1 and not visited[row - 1][col]:
      if levels[row - 1][col] == 1:
        count += 1
      else:
        queue.insert(0, (row - 1, col))
        visited[row - 1][col] = True
    if row + 1 < r + 2 and not visited[row + 1][col]:
      if levels[row + 1][col] == 1:
        count += 1
      else:
        queue.insert(0, (row + 1, col))
        visited[row + 1][col] = True
    if col - 1 > -1 and not visited[row][col - 1]:
      if levels[row][col - 1] == 1:
        count += 1
      else:
        queue.insert(0, (row, col - 1))
        visited[row][col - 1] = True
    if col + 1< c + 2 and not visited[row][col + 1]:
      if levels[row][col + 1] == 1:
        count += 1
      else:
        queue.insert(0, (row, col + 1))
        visited[row][col + 1] = True
  return count


def main():
  r, c = map(int, stdin.readline().split())
  levels = list()
  visited = list()

  levels.append([0] * (c + 2))
  visited.append([False] * (c + 2))
  for x in range(r):
    add = [0]
    add.extend([int(y) for y in list(stdin.readline().strip())])
    add.append(0)

    levels.append(add)
    visited.append([False] * (c + 2))
  visited.append([False] * (c + 2))
  levels.append([0] * (c + 2))

  print(bfs(levels, r, c, visited))


if __name__ == "__main__":
  main()

File: JAVA/coast/coast.java
// Inputs: n, m, n rows of signals
// Outputs: count

import java.util.LinkedList;
import java.util.Scanner;

public class coast {
  public static void main(String[] args) {
    Scanner in = new Scanner(System.in);

    String[] n_m = in.nextLine().split(" ");
    int n = Integer.parseInt(n_m[0]);
    int m = Integer.parseInt(n_m[1]);

    int[][] map = new int[n + 2][m + 2];
    boolean[][] visited = new boolean[n + 2][m + 2];

    for (int r = 1; r < n + 1; r++) {
      String[] row = in.nextLine().split("");
      for (int c = 1; c < m + 1; c++) {
        map[r][c] = Integer.parseInt(row[c - 1]);
      }
    }

    int count = bfs(map, visited, n, m);
    System.out.println(count);
  }

  public static int bfs(int[][] map, boolean[][] visited, int n, int m) {
    LinkedList<Point> q = new LinkedList<Point>();
    q.add(new Point(0, 0));
    visited[0][0] = true;
    int count = 0;

    while (!q.isEmpty()) {
      Point curr = q.poll();

      if (curr.r - 1 > -1 && !visited[curr.r - 1][curr.c]) {
        if (map[curr.r - 1][curr.c] == 1) {
          count += 1;
        } else {
          q.add(new Point(curr.r - 1, curr.c));
          visited[curr.r - 1][curr.c] = true;
        }
      }
      if (curr.r + 1 < n + 2 && !visited[curr.r + 1][curr.c]) {
        if (map[curr.r + 1][curr.c] == 1) {
          count += 1;
        } else {
          q.add(new Point(curr.r + 1, curr.c));
          visited[curr.r + 1][curr.c] = true;
        }
      }
      if (curr.c - 1 > -1 && !visited[curr.r][curr.c - 1]) {
        if (map[curr.r][curr.c - 1] == 1) {
          count += 1;
        } else {
          q.add(new Point(curr.r, curr.c - 1));
          visited[curr.r][curr.c - 1] = true;
        }
      }
      if (curr.c + 1 < m + 2 && !visited[curr.r][curr.c + 1]) {
        if (map[curr.r][curr.c + 1] == 1) {
          count += 1;
        } else {
          q.add(new Point(curr.r, curr.c + 1));
          visited[curr.r][curr.c + 1] = true;
        }
      }
    }

    return count;
  }

  static class Point {
    int r = 0;
    int c = 0;

    Point(int r, int c) {
      this.r = r;
      this.c = c;
    }
  }
}
